README.md
# SIESTAstepper
[![PyPI version](https://badge.fury.io/py/SIESTAstepper.svg)](https://badge.fury.io/py/SIESTAstepper)
[![PyPI pyversions](https://img.shields.io/pypi/pyversions/SIESTAstepper.svg)](https://pypi.python.org/pypi/SIESTAstepper/)
[![Python package](https://github.com/eftalgezer/SIESTAstepper/actions/workflows/python-package.yml/badge.svg)](https://github.com/eftalgezer/SIESTAstepper/actions/workflows/python-package.yml)
[![codecov](https://codecov.io/gh/eftalgezer/SIESTAstepper/branch/main/graph/badge.svg?token=Q9TJFIN1U1)](https://codecov.io/gh/eftalgezer/SIESTAstepper)
[![Codacy Badge](https://app.codacy.com/project/badge/Coverage/4ff526bd45e642bb81b300f2243baef2)](https://www.codacy.com/gh/eftalgezer/SIESTAstepper/dashboard?utm_source=github.com&utm_medium=referral&utm_content=eftalgezer/SIESTAstepper&utm_campaign=Badge_Coverage)
[![PyPI download month](https://img.shields.io/pypi/dm/SIESTAstepper.svg)](https://pypi.python.org/pypi/SIESTAstepper/)
[![PyPI download week](https://img.shields.io/pypi/dw/SIESTAstepper.svg)](https://pypi.python.org/pypi/SIESTAstepper/)
[![PyPI download day](https://img.shields.io/pypi/dd/SIESTAstepper.svg)](https://pypi.python.org/pypi/SIESTAstepper/)
![GitHub all releases](https://img.shields.io/github/downloads/eftalgezer/SIESTAstepper/total?style=flat)
[![GitHub contributors](https://img.shields.io/github/contributors/eftalgezer/SIESTAstepper.svg)](https://github.com/eftalgezer/SIESTAstepper/graphs/contributors/)
[![CodeFactor](https://www.codefactor.io/repository/github/eftalgezer/siestastepper/badge)](https://www.codefactor.io/repository/github/eftalgezer/siestastepper)
[![Codacy Badge](https://app.codacy.com/project/badge/Grade/4ff526bd45e642bb81b300f2243baef2)](https://www.codacy.com/gh/eftalgezer/SIESTAstepper/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=eftalgezer/SIESTAstepper&amp;utm_campaign=Badge_Grade)
[![PyPI license](https://img.shields.io/pypi/l/SIESTAstepper.svg)](https://pypi.python.org/pypi/SIESTAstepper/)
[![DOI](https://zenodo.org/badge/532944393.svg)](https://zenodo.org/badge/latestdoi/532944393)

SIESTAstepper runs SIESTA step by step, designed for constrained calculations.

## Installation

Use the package manager [pip](https://pip.pypa.io/en/stable/) to install SIESTAstepper.

```bash
$ pip install SIESTAstepper

# to make sure you have the latest version
$ pip install -U SIESTAstepper

# latest available code base
$ pip install -U git+https://github.com/eftalgezer/SIESTAstepper.git
```

## Tutorial

- [SIESTAstepper v1.0.0 tutorial](https://beyondthearistotelian.blogspot.com/2022/09/siestastepper-v100-tutorial.html)
- [SIESTAstepper v1.1.0 tutorial](https://beyondthearistotelian.blogspot.com/2022/09/siestastepper-v110-tutorial.html)
- [What is new in SIESTAstepper v1.2.0?](https://beyondthearistotelian.blogspot.com/2022/09/what-is-new-in-siestastepper-v120.html)
- [SIESTAstepper v2.0.0 tutorial](https://beyondthearistotelian.blogspot.com/2022/09/siestastepper-v200-tutorial.html)

## Usage

All SIESTA working directories must be named as i1, i2, i3 ... and so on.

### In code

#### Simple usage

```python
SIESTAstepper.run("graphene")

```

#### Advance usage

```python
import SIESTAstepper

# Sets the path of the working directory
SIESTAstepper.settings.set_cwd("path/to/working/directory")

# Sets the name of SIESTA log files (default is "log")
SIESTAstepper.settings.set_log("log")

# Sets number of cores for parallel run
SIESTAstepper.settings.set_cores(4)

# Sets Anaconda environment
SIESTAstepper.settings.set_conda("envir")

# Sets the subfolder name for interrupted calculations (default is "continue")
SIESTAstepper.settings.set_cont("continue")

# Sets the SIESTA command name (default is "siesta")
SIESTAstepper.settings.set_siesta("siesta_p")

#Sets the file type to continue next step (default is "log", possible values are "log", "XV", "ANI")
SIESTAstepper.settings.set_contfrom("ANI")

# Sets the filenames to copy (useful for interrupted calculations)
SIESTAstepper.settings.contfiles.extend(["file1", "file2"])

# Sets the extensions to copy (useful for interrupted calculations, default is ["psf, "fdf"]
SIESTAstepper.settings.contextensions.extend(["DM", "XV", "CG", "LWF"])

# Runs SIESTA step by step
SIESTAstepper.run("graphene")

# Converts last geometry of an ANI to FDF by using the previous FDF and ANI files
SIESTAstepper.ani_to_fdf("path/to/ANI", "path/to/FDF", "path/to/newFDF")

# Converts XYZ to FDF by using the previous FDF and XYZ files
SIESTAstepper.xyz_to_fdf("path/to/XYZ", "path/to/FDF", "path/to/newFDF")

# Merges ANI files
SIESTAstepper.merge_ani(label = "graphene")

# Merges ANI files by setting a path
SIESTAstepper.merge_ani(label = "graphene", path = "path/to/i*/ANI/files")

# Creates GIF file from a given ANI
SIESTAstepper.ani_to_gif(anifile="graphene.ANI")

# Creates GIF file from a given ANI by setting width and height (defaults are 1920 and 1080, respectively)
SIESTAstepper.ani_to_gif(anifile="graphene.ANI", width=1920, height=1080)

# Creates GIF file from a given ANI by setting the bonds between atoms (default is 1.3)
SIESTAstepper.ani_to_gif(anifile="graphene.ANI", bonds_param=1.3)

# Creates GIF file from a given ANI by setting loop (default is 0; 0 means loop, 1 means no loop)
SIESTAstepper.ani_to_gif(anifile="graphene.ANI", loop=1)

# Creates GIF file from a given ANI by setting camera
SIESTAstepper.ani_to_gif(anifile="graphene.ANI", camera=((40, 0, 0),
                                                        (0, 0, 0),
                                                        (0, 1, 0)))

# Runs SIESTA for a given step
SIESTAstepper.run_next("1", "graphene")

# Runs SIESTA for a given step without continuing to next step
SIESTAstepper.single_run("1", "graphene")

# Continues to an interrupted calculation without continuing to next step
SIESTAstepper.single_run_interrupted("1", "graphene")

# Continues to an interrupted calculation
SIESTAstepper.run_interrupted("1", "graphene")

# Creates folders named i1, i2, i3, i4, i5, ...
SIESTAstepper.make_directories(5)

# Copies files from i1 to i2
SIESTAstepper.copy_files(["psf", "fdf", "XV", "DM"], "graphene", "path/to/i1", "path/to/i2")

# Plots and returns energies from log files
SIESTAstepper.energy_analysis()

# Returns energies from log files without plotting
SIESTAstepper.energy_analysis(plot_ = False)

# Plots and returns energies from log files by setting a path
SIESTAstepper.energy_analysis(path = "path/to/i*/log/files")

# Calculates the energy differences between minima and maxima
SIESTAstepper.energy_diff()

# Calculates the energy differences between minima and maxima by setting a path
SIESTAstepper.energy_diff(path = "path/to/i*/log/files")

# Plots and returns forces from log files (defaults are "Tot" and "atomic", respectively)
SIESTAstepper.force_analysis(atomindex="Tot", forcetype="atomic")

# Returns forces from log files without plotting
SIESTAstepper.force_analysis(plot_ = False)

# Plots and returns forces from log files by setting a path
SIESTAstepper.force_analysis(path = "path/to/i*/log/files")

# Calculates the force differences between minima and maxima (defaults are "Tot" and "atomic", respectively)
SIESTAstepper.force_diff(atomindex="Tot", forcetype="atomic")

# Calculates the force differences between minima and maxima by setting a path
SIESTAstepper.force_diff(path = "path/to/i*/log/files")

# Calculates the pair correlation function
SIESTAstepper.pair_correlation_function(label="graphene")

# Calculates the pair correlation function without plotting
SIESTAstepper.pair_correlation_function(label="graphene", plot_=True)

# Calculates the pair correlation function by setting a path
SIESTAstepper.pair_correlation_function(label="graphene", path="path/to/i*")

# Calculates the pair correlation function by setting infinitesimal dr parameter
SIESTAstepper.pair_correlation_function(label="graphene", dr=0.1)
```

### In terminal

#### Simple usage

```sh
$ python -m SIESTAstepper run log graphene
```

#### Advance usage

```sh
$ python -m SIESTAstepper run log graphene

$ python -m SIESTAstepper run log graphene contfiles=file1,file2 contextensions=DM,XV,CG,LWF

$ python -m SIESTAstepper run log graphene mpirun=4

$ python -m SIESTAstepper run log graphene conda=envir

$ python -m SIESTAstepper run log graphene siesta=siesta_p

$ python -m SIESTAstepper run_next log 1 graphene

$ python -m SIESTAstepper run_next log 1 graphene contfiles=file1,file2 contextensions=DM,XV,CG,LWF

$ python -m SIESTAstepper run_next log 1 graphene mpirun=4

$ python -m SIESTAstepper run_next log 1 graphene conda=envir

$ python -m SIESTAstepper run_next log 1 graphene siesta=siesta_p

$ python -m SIESTAstepper single_run log 1 graphene

$ python -m SIESTAstepper single_run log 1 graphene mpirun=4

$ python -m SIESTAstepper single_run log 1 graphene conda=envir

$ python -m SIESTAstepper single_run log 1 graphene siesta=siesta_p

$ python -m SIESTAstepper run_interrupted log 1 graphene cont=continue

$ python -m SIESTAstepper run_interrupted log 1 graphene cont=continue contfiles=file1,file2 contextensions=DM,XV,CG,LWF

$ python -m SIESTAstepper run_interrupted log 1 graphene mpirun=4 cont=continue

$ python -m SIESTAstepper run_interrupted log 1 graphene conda=envir cont=continue

$ python -m SIESTAstepper run_interrupted log 1 graphene siesta=siesta_p cont=continue

$ python -m SIESTAstepper single_run_interrupted log 1 graphene cont=continue

$ python -m SIESTAstepper single_run_interrupted log 1 graphene mpirun=4 cont=continue

$ python -m SIESTAstepper single_run_interrupted log 1 graphene conda=envir cont=continue

$ python -m SIESTAstepper single_run_interrupted log 1 graphene siesta=siesta_p cont=continue

$ python -m SIESTAstepper make_directories 5

$ python -m SIESTAstepper copy_files graphene path/to/i1 path/to/i2 psf fdf XV DM

$ python -m SIESTAstepper copy_files graphene path/to/i1 path/to/i2 psf fdf XV DM

$ python -m SIESTAstepper ani_to_fdf path/to/ANI path/to/FDF path/to/newFDF

$ python -m SIESTAstepper xyz_to_fdf path/to/XYZ path/to/FDF path/to/newFDF

$ python -m SIESTAstepper merge_ani graphene

$ python -m SIESTAstepper merge_ani graphene cont=continue

$ python -m SIESTAstepper merge_ani graphene path=path/to/i*/ANI/files

$ python -m SIESTAstepper ani_to_gif graphene.ANI 1920 1080 1 1.3

$ python -m SIESTAstepper ani_to_gif graphene.ANI 1920 1080 1 1.3 camera=40,0,0,0,0,0,0,1,0

$ python -m SIESTAstepper energy_analysis log

$ python -m SIESTAstepper energy_analysis log cont=continue

$ python -m SIESTAstepper energy_analysis log noplot

$ python -m SIESTAstepper energy_analysis log path=path/to/i*/log/files

$ python -m SIESTAstepper energy_diff log

$ python -m SIESTAstepper energy_diff log cont=continue

$ python -m SIESTAstepper energy_diff log path=path/to/i*/log/files

$ python -m SIESTAstepper force_analysis log atomic Tot

$ python -m SIESTAstepper force_analysis log atomic Tot cont=continue

$ python -m SIESTAstepper force_analysis log atomic Tot noplot

$ python -m SIESTAstepper force_analysis log atomic Tot path=path/to/i*/log/files

$ python -m SIESTAstepper force_diff log atomic Tot

$ python -m SIESTAstepper energy_diff log atomic Tot cont=continue

$ python -m SIESTAstepper energy_diff log atomic Tot path=path/to/i*/log/files

```

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## Citation
If you are using SIESTAstepper, please citate relevant version. You can find the relevant citation [here](https://doi.org/10.5281/zenodo.7051271).

```bibtex
@software{eftal_gezer_2022_7212472,
  author       = {Eftal Gezer},
  title        = {eftalgezer/SIESTAstepper: v2.1.0},
  month        = oct,
  year         = 2022,
  publisher    = {Zenodo},
  version      = {v2.1.0},
  doi          = {10.5281/zenodo.7212472},
  url          = {https://doi.org/10.5281/zenodo.7212472}
}
```

## License
[GNU General Public License v3.0](https://github.com/eftalgezer/SIESTAstepper/blob/master/LICENSE) 

## Acknowledgments

`ani_to_gif` function uses [ANIAnimator](https://github.com/eftalgezer/ANIAnimator).

## References

Eftal Gezer. (2022). eftalgezer/ANIAnimator: v0.2.1 (v0.2.1). Zenodo. https://doi.org/10.5281/zenodo.7182193


CHANGELOG.md
# Changelog

## [v2.1.0](https://github.com/eftalgezer/SIESTAstepper/tree/v2.1.0) (2022-10-16)

- Backwards compatibility
- `analysis` is renamed as `energy_analysis`
- `force_analysis`, `force_diff`, `pair_correlation_function`, and `ani_to_gif` functions are added.

**Full Changelog**: https://github.com/eftalgezer/SIESTAstepper/compare/v2.0.0...v2.1.0

Eftal Gezer. (2022). eftalgezer/ANIAnimator: v0.2.1 (v0.2.1). Zenodo. https://doi.org/10.5281/zenodo.7182193

## [v2.0.0](https://github.com/eftalgezer/SIESTAstepper/tree/v2.0.0) (2022-09-23)

- Bug fix
- Linting
- `update_cwd`, `update_log`, `update_cores`, `update_conda`, `update_cont`, `update_siesta`, `contfiles`, ND `contextensions` are moved in `SIESTAstepper.settings` and renamed. You can reach those by `SIESTAstepper.settings.set_cwd`, `SIESTAstepper.settings.set_log`, `SIESTAstepper.settings.set_cores`, `SIESTAstepper.settings.set_conda`, `SIESTAstepper.settings.set_cont`, `SIESTAstepper.settings.set_siesta`, `SIESTAstepper.settings.contfiles`, and `SIESTAstepper.settings.contextensions`, respectively.
- `SIESTAstepper.settings.get_cwd`, `SIESTAstepper.settings.get_log`, `SIESTAstepper.settings.get_cores`, `SIESTAstepper.settings.get_conda`, `SIESTAstepper.settings.get_cont`, and `SIESTAstepper.settings.get_siesta` are added.
- The previous version was checking the restart parameters and the `Userbasis` parameter in a fdf file by `.true.` and `.false.`. `T` and `F` support is added.
- More tests are added.
- [A tutorial for SIESTAstepper v2.0.0](https://beyondthearistotelian.blogspot.com/2022/09/siestastepper-v200-tutorial.html) is published.

**Full Changelog**: https://github.com/eftalgezer/SIESTAstepper/compare/v1.2.2...v2.0.0

## [v1.2.3](https://github.com/eftalgezer/SIESTAstepper/tree/v1.2.3) (2022-09-19)

Bug fix

**Full Changelog**: https://github.com/eftalgezer/SIESTAstepper/compare/v1.2.2...v1.2.3

## [v1.2.2](https://github.com/eftalgezer/SIESTAstepper/tree/v1.2.2) (2022-09-19)

- Tests added
- Bug fix

**Full Changelog**: https://github.com/eftalgezer/SIESTAstepper/compare/v1.2.1...v1.2.2

## [v1.2.1](https://github.com/eftalgezer/SIESTAstepper/tree/v1.2.1) (2022-09-16)

Bug fix

**Full Changelog**: https://github.com/eftalgezer/SIESTAstepper/compare/v1.2.0...v1.2.1

## [v1.2.0](https://github.com/eftalgezer/SIESTAstepper/tree/v1.2.0) (2022-09-16)

- CG and LWF support to continue interrupted calculations
- ion support as a pseudopotential file
- `update_siesta` function is added. It is possible to change the SIESTA run command.

For details, please see [What is new in SIESTAstepper v1.2.0?](https://beyondthearistotelian.blogspot.com/2022/09/what-is-new-in-siestastepper-v120.html)

**Full Changelog**: https://github.com/eftalgezer/SIESTAstepper/compare/v1.1.0...v1.2.0

## [v1.1.0](https://github.com/eftalgezer/SIESTAstepper/tree/v1.1.0) (2022-09-15)

- `energy_diff` function is added. It is possible to get energy differences between minima and maxima.
- Bug fixes
- - The bug on the file name sort is fixed.
- - `update_cwd`, `update_log`, `update_cores`, `update_conda`, and `update_cont` functions are added to change the values of `cwd`, `log`, `cores`, `conda`, and `cont`, respectively.
- [A tutorial for v1.1.0 is written](https://beyondthearistotelian.blogspot.com/2022/09/siestastepper-v110-tutorial.html)

**Full Changelog**: https://github.com/eftalgezer/SIESTAstepper/compare/v1.0.0...v1.1.0

## [v1.0.0](https://github.com/eftalgezer/SIESTAstepper/tree/v1.0.0) (2022-09-14)

- `run_interrupted` function is added. It is possible to continue from an interrupted calculation.
- `make_directories` function is added. It is possible to make i* directories massively. 
- Security improvements
- Setting the NumberOfAtoms parameter while creating an fdf file
- `copy_files` function is added. It is possible to copy desired files by extension.
- Print PID on screen
- `contfiles` parameter is added. It is used in the copy_files function to copy files, not by extension.
- Continuations are infinitive. If you kill your job in the "continue" folder, you can continue with the "continue_2", "continue_3" folder, etc.
- Parameter `missing` is removed in the functions `merge_ani`, and `analysis`. These functions will advantage the benefit of the `SIESTAstepper.cont` parameter.
- Improved Windows support
- `single_run` function is added. It is possible to run a step without proceeding to the next step.
- `single_run_interrupted` function is added. It is possible to run an interrupted step without proceeding to the next step.
- CHANGELOG.md is created.
- [A tutorial for v1.0.0 is written.](https://beyondthearistotelian.blogspot.com/2022/09/siestastepper-v100-tutorial.html)
- Bug fixes

Thanks to [toppk](Bug fix) at Stackoverflow for the [help in fixing the bug in the `remove_nones` function](https://stackoverflow.com/questions/73721062/why-the-operator-does-not-work-as-expected-in-python).

**Full Changelog**: https://github.com/eftalgezer/SIESTAstepper/compare/v0.4.2...v1.0.0

## [v0.4.2](https://github.com/eftalgezer/SIESTAstepper/tree/v0.4.2) (2022-09-08)

[Full Changelog](https://github.com/eftalgezer/SIESTAstepper/compare/v0.4.1...v0.4.2)

## [v0.4.1](https://github.com/eftalgezer/SIESTAstepper/tree/v0.4.1) (2022-09-08)

[Full Changelog](https://github.com/eftalgezer/SIESTAstepper/compare/v0.4.0...v0.4.1)

## [v0.4.0](https://github.com/eftalgezer/SIESTAstepper/tree/v0.4.0) (2022-09-07)

[Full Changelog](https://github.com/eftalgezer/SIESTAstepper/compare/v0.3.0...v0.4.0)

## [v0.3.0](https://github.com/eftalgezer/SIESTAstepper/tree/v0.3.0) (2022-09-06)

[Full Changelog](https://github.com/eftalgezer/SIESTAstepper/compare/v0.2.0...v0.3.0)

## [v0.2.0](https://github.com/eftalgezer/SIESTAstepper/tree/v0.2.0) (2022-09-06)

[Full Changelog](https://github.com/eftalgezer/SIESTAstepper/compare/v0.1.0...v0.2.0)

## [v0.1.0](https://github.com/eftalgezer/SIESTAstepper/tree/v0.1.0) (2022-09-05)


SIESTAstepper/__main__.py
"""
SIESTAstepper terminal client
"""
from __future__ import absolute_import
import sys
from .core import (
    run,
    single_run,
    run_next,
    run_interrupted,
    single_run_interrupted,
    make_directories,
    copy_files,
    ani_to_fdf,
    xyz_to_fdf,
    xv_to_fdf,
    log_to_fdf,
    xv_to_ani,
    merge_ani,
    ani_to_gif,
    energy_analysis,
    force_analysis,
    energy_diff,
    force_diff,
    pair_correlation_function,
    settings
)


def main(args):
    """Main function"""
    function = args[1]
    for arg in args:
        independents(arg)
    if function not in ["run", "single_run", "run_next", "run_interrupted",
                        "single_run_interrupted", "make_directories",
                        "copy_files", "ani_to_fdf", "xyz_to_fdf", "xv_to_fdf",
                        "log_to_fdf", "xv_to_ani", "merge_ani", "ani_to_gif",
                        "energy_analysis", "force_analysis", "energy_diff",
                        "force_diff", "pair_correlation_function"]:
        raise AttributeError(
            """Command not found. Please use 'run',
            'single_run', 'run_next', 'run_interrupted',
            'single_run_interrupted', 'make_directories',
            'copy_files', 'ani_to_fdf', 'xyz_to_fdf',
            'xv_to_fdf', 'log_to_fdf', 'xv_to_ani',
            'merge_ani', 'ani_to_gif', 'energy_analysis',
            'force_analysis', 'energy_diff', 'force_diff',
            'pair_correlation_function'""".replace("           ", "").replace("\n", "")
        )
    if function == "run":
        settings.set_log(args[2])
        run(args[3])
    elif function == "single_run":
        settings.set_log(args[2])
        single_run(args[3], args[4])
    elif function == "run_next":
        settings.set_log(args[2])
        run_next(args[3], args[4])
    elif function == "run_interrupted":
        settings.set_log(args[2])
        run_interrupted(args[3], args[4])
    elif function == "single_run_interrupted":
        settings.set_log(args[2])
        single_run_interrupted(args[3], args[4])
    elif function == "make_directories":
        make_directories(int(args[2]))
    elif function == "copy_files":
        copy_files(
            [_ for _ in args[5:] if not _.startswith("contfiles=")],
            args[2],
            args[3],
            args[4]
        )
    elif function == "ani_to_fdf":
        ani_to_fdf(args[2], args[3], args[4])
    elif function == "xyz_to_fdf":
        xyz_to_fdf(args[2], args[3], args[4])
    elif function == "xv_to_fdf":
        xv_to_fdf(args[2], args[3], args[4])
    elif function == "log_to_fdf":
        log_to_fdf(args[2], args[3], args[4])
    elif function == "xv_to_ani":
        if len(args) == 3:
            xv_to_ani(label=args[2])
        elif len(args) > 3:
            for arg in args[3:]:
                if arg.startswith("path="):
                    path = arg.split("=")[1]
            xv_to_ani(label=args[2], path=path)
    elif function == "merge_ani":
        path = "i*"
        if len(args) == 3:
            merge_ani(label=args[2])
        elif len(args) > 3:
            for arg in args[3:]:
                if arg.startswith("path="):
                    path = arg.split("=")[1]
            merge_ani(label=args[2], path=path)
    elif function == "ani_to_gif":
        width = None
        height = None
        loop = None
        bonds_param = None
        camera = None
        try:
            width = args[3]
        except IndexError:
            pass
        try:
            height = args[4]
        except IndexError:
            pass
        try:
            loop = args[5]
        except IndexError:
            pass
        try:
            bonds_param = args[6]
        except IndexError:
            pass
        try:
            camera = args[7]
        except IndexError:
            pass
        if width is None:
            width = 1920
        if height is None:
            height = 1080
        if loop is None:
            loop = 1
        if bonds_param is None:
            bonds_param = 1.3
        if camera == "default":
            camera = None
        elif camera is not None:
            cameraparams = camera.split(",")
            camera = (
                (int(cameraparams[0]), int(cameraparams[1]), int(cameraparams[2])),
                (int(cameraparams[3]), int(cameraparams[4]), int(cameraparams[5])),
                (int(cameraparams[6]), int(cameraparams[7]), int(cameraparams[8]))
            )
        ani_to_gif(args[3], width, height, loop, bonds_param, camera)
    elif function == "energy_analysis":
        settings.set_log(args[2])
        plot_ = True
        path = "i*"
        if len(args) > 5:
            for arg in args[4:]:
                if arg.startswith("path="):
                    path = arg.split("=")[1]
                if arg == "noplot":
                    plot_ = False
            energy_analysis(energytype=args[3], path=path, plot_=plot_)
        else:
            energy_analysis(energytype=args[3])
    elif function == "force_analysis":
        settings.set_log(args[2])
        plot_ = True
        path = "i*"
        if len(args) > 6:
            for arg in args[5:]:
                if arg.startswith("path="):
                    path = arg.split("=")[1]
                if arg == "noplot":
                    plot_ = False
            force_analysis(atomindex=args[4], forcetype=args[3], path=path, plot_=plot_)
        else:
            force_analysis(atomindex=args[4], forcetype=args[3])
    elif function == "energy_diff":
        settings.set_log(args[2])
        path = "i*"
        if len(args) > 5:
            for arg in args[4:]:
                if arg.startswith("path="):
                    path = arg.split("=")[1]
            energy_diff(energytype=args[3], path=path)
        else:
            energy_diff(energytype=args[3])
    elif function == "force_diff":
        settings.set_log(args[2])
        path = "i*"
        if len(args) > 6:
            for arg in args[5:]:
                if arg.startswith("path="):
                    path = arg.split("=")[1]
            force_diff(atomindex=args[4], forcetype=args[3], path=path)
        else:
            force_diff(atomindex=args[4], forcetype=args[3])
    elif function == "pair_correlation_function":
        path = "i*"
        dr = 0.1
        plot_ = True
        if len(args) > 4:
            for arg in args[3:]:
                if arg.startswith("path="):
                    path = arg.split("=")[1]
                if arg.startswith("dr="):
                    dr = arg.split("=")[1]
                if arg == "noplot":
                    plot_ = False
            pair_correlation_function(label=args[2], path=path, dr=dr, plot_=plot_)
        else:
            pair_correlation_function(label=args[2])


def independents(arg):
    """Sets independent variables"""
    if arg.startswith("mpirun="):
        settings.set_cores(int(arg.split("=")[1]))
    if arg.startswith("conda="):
        settings.set_conda(arg.split("=")[1])
    if arg.startswith("cont="):
        settings.set_cont(arg.split("=")[1])
    if arg.startswith("contfrom="):
        settings.set_contfrom(arg.split("=")[1])
    if arg.startswith("contfiles="):
        settings.contfiles.extend(arg.split("=")[1].split(","))
    if arg.startswith("contextensions="):
        settings.contextensions.extend(arg.split("=")[1].split(","))
    if arg.startswith("siesta="):
        settings.set_siesta(arg.split("=")[1])


if __name__ == "__main__":
    main(args=sys.argv)
